build.gradle

```diff
@@ -77,6 +79,37 @@ subprojects {
         }
     }
 
+    plugins.withId('maven-publish') {
+        publishing {
+            publications {
+                nebula(MavenPublication)
+            }
+            repositories {
+                maven {
+                    name = 'Milestone'
+                    url = 'https://repo.spring.io/milestone'
+                    credentials {
+                        username findProperty('MILESTONE_REPO_USER')
+                        password findProperty('MILESTONE_REPO_PASSWORD')
+                    }
+                }
+                maven {
+                    name = 'MavenCentral'
+                    url = 'https://s01.oss.sonatype.org/service/local/staging/deploy/maven2'
+                    credentials {
+                        username findProperty('MAVEN_CENTRAL_USER')
+                        password findProperty('MAVEN_CENTRAL_PASSWORD')
+                    }
+                }
+            }
+        }
+
+        signing {
+            useInMemoryPgpKeys(findProperty('SIGNING_KEY'), findProperty('SIGNING_PASSWORD'))
+            sign publishing.publications.nebula
```

Comment:
Will our snapshot releases be signed too? I wasn't sure given the setup.

Reply:
I tried to do a `pTML` (publishToMavenLocal) which is what I do to test changes in snapshots locally, but it failed because the signing properties aren't configured. We want contributors to be able to publish local snapshots without needing to configure signing keys. That said, we do want to ensure signing is done for remote publishing. I'm not sure the best way to go about that. I added
```
if (hasProperty('SIGNING_KEY') && hasProperty('SIGNING_PASSWORD')) {
```
but that makes it possible for remote publishing to be attempted without signing artifacts, which I think we want to avoid - it will fail Maven Central's requirements, for example, and is just good practice if we easily can.

Reply:
Oops, I messed this up. This signs snapshots too as well as local publications.
I fixed it this way:
```groovy
signing {
            required = hasProperty('SIGNING_KEY')
}
```

This tells the signing plugin that signature is required when the key is present otherwise the task will be skipped so if you have the key it will sign local builds, if you don't it won't. I did not include the password since the gpg key can be passwordless.

Reply:
That fixes the issue of contributors being able to publish a snapshot to their local maven repository. We should probably still think about checking that signing is configured if we're publishing a release to a remote repository.